File: backend/services/deezer_service.py
import httpx
import os
from typing import List, Dict, Any, Optional
from dotenv import load_dotenv

class DeezerService:
    def __init__(self):
        # Load environment variables from parent directory
        load_dotenv(dotenv_path="../.env")
        self.base_url = "https://api.deezer.com"
    
    async def search_tracks(self, query: str, limit: int = 20) -> List[Dict[str, Any]]:
        """Search for tracks on Deezer"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/search/track",
                    params={"q": query, "limit": limit}
                )
                response.raise_for_status()
                data = response.json()
                
                tracks = []
                for track in data.get("data", []):
                    formatted_track = {
                        "id": str(track["id"]),
                        "title": track["title"],
                        "artist": {
                            "id": str(track["artist"]["id"]),
                            "name": track["artist"]["name"]
                        },
                        "album": {
                            "id": str(track["album"]["id"]),
                            "title": track["album"]["title"],
                            "cover_medium": track["album"]["cover_medium"]
                        },
                        "preview": track.get("preview"),
                        "duration": track["duration"],
                        "link": track["link"]
                    }
                    tracks.append(formatted_track)
                
                return tracks
        except Exception as e:
            print(f"Error searching Deezer tracks: {e}")
            return []
    
    async def get_track(self, track_id: str) -> Optional[Dict[str, Any]]:
        """Get track details from Deezer"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(f"{self.base_url}/track/{track_id}")
                response.raise_for_status()
                track = response.json()
                
                return {
                    "id": str(track["id"]),
                    "title": track["title"],
                    "artist": {
                        "id": str(track["artist"]["id"]),
                        "name": track["artist"]["name"]
                    },
                    "album": {
                        "id": str(track["album"]["id"]),
                        "title": track["album"]["title"],
                        "cover_medium": track["album"]["cover_medium"]
                    },
                    "preview": track.get("preview"),
                    "duration": track["duration"],
                    "link": track["link"]
                }
        except Exception as e:
            print(f"Error getting Deezer track {track_id}: {e}")
            return None
    
    async def get_preview_url(self, track_id: str) -> Optional[str]:
        """Get preview URL for a track"""
        try:
            track = await self.get_track(track_id)
            return track.get("preview_url") if track else None
        except Exception as e:
            print(f"Error getting preview URL for track {track_id}: {e}")
            return None
    
    async def search_tracks_by_spotify_info(self, spotify_track: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Search for a track on Deezer using Spotify track info"""
        try:
            # Create search query from Spotify track info
            artist_name = spotify_track.get("artists", [{}])[0].get("name", "")
            track_name = spotify_track.get("name", "")
            
            if not artist_name or not track_name:
                return None
            
            # Search with artist and track name
            query = f'artist:"{artist_name}" track:"{track_name}"'
            tracks = await self.search_tracks(query, 5)
            
            if not tracks:
                # Fallback to simple search
                query = f"{artist_name} {track_name}"
                tracks = await self.search_tracks(query, 5)
            
            # Return the first match
            return tracks[0] if tracks else None
            
        except Exception as e:
            print(f"Error searching Deezer for Spotify track: {e}")
            return None

File: src/services/deezerService.ts
// src/services/deezerService.ts
import { SpotifyTrack } from "../types/music";

export interface DeezerTrack {
  id: string;
  title: string;
  artist: {
    id: string;
    name: string;
  };
  album: {
    id: string;
    title: string;
    cover_medium: string;
  };
  duration: number;
  preview: string; // This is the preview URL
  link: string;
}

export interface DeezerSearchResponse {
  data: DeezerTrack[];
  total: number;
}

class DeezerService {
  private readonly BASE_URL = "https://api.deezer.com";

  /**
   * Search for tracks on Deezer
   */
  async searchTracks(query: string, limit: number = 20): Promise<DeezerTrack[]> {
    try {
      const encodedQuery = encodeURIComponent(query);
      const url = `${this.BASE_URL}/search/track?q=${encodedQuery}&limit=${limit}`;
      
      console.log("🎵 Searching Deezer for:", query);
      
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`Deezer API error: ${response.status}`);
      }
      
      const data: DeezerSearchResponse = await response.json();
      console.log(`✅ Found ${data.data.length} tracks on Deezer`);
      
      return data.data;
    } catch (error) {
      console.error("❌ Deezer search failed:", error);
      return [];
    }
  }

  /**
   * Search for tracks by artist
   */
  async searchTracksByArtist(artistName: string, limit: number = 20): Promise<DeezerTrack[]> {
    return this.searchTracks(`artist:"${artistName}"`, limit);
  }

  /**
   * Get recommendations based on Spotify track
   */
  async getRecommendationsForSpotifyTrack(spotifyTrack: SpotifyTrack): Promise<DeezerTrack[]> {
    // Search for the same track on Deezer
    const query = `${spotifyTrack.name} ${spotifyTrack.artists[0]?.name}`;
    return this.searchTracks(query, 10);
  }

  /**
   * Get recommendations for a vibe mode
   */
  async getRecommendationsForVibeMode(vibeMode: string): Promise<DeezerTrack[]> {
    const searchTerms = [
      `${vibeMode} music`,
      `popular ${vibeMode}`,
      `${vibeMode} hits`,
      `trending ${vibeMode}`
    ];

    const allTracks: DeezerTrack[] = [];
    
    for (const term of searchTerms) {
      try {
        const tracks = await this.searchTracks(term, 15);
        allTracks.push(...tracks);
      } catch (err) {
        console.warn(`Failed to search Deezer for "${term}":`, err);
      }
    }

    // Remove duplicates by ID first, then by name
    const uniqueByIdTracks = allTracks.filter((track, index, self) => 
      index === self.findIndex(t => t.id === track.id) && track.preview
    );
    
    // Remove duplicates by name (case-insensitive)
    const seenNames = new Set<string>();
    const uniqueTracks = uniqueByIdTracks.filter(track => {
      const normalizedName = track.title.toLowerCase().trim();
      if (!seenNames.has(normalizedName)) {
        seenNames.add(normalizedName);
        return true;
      }
      return false;
    });

    console.log(`🎵 Found ${uniqueTracks.length} unique tracks with previews on Deezer`);
    return uniqueTracks.slice(0, 50);
  }

  /**
   * Convert Deezer track to Spotify-like format for compatibility
   */
  convertToSpotifyFormat(deezerTrack: DeezerTrack): SpotifyTrack {
    return {
      id: `deezer_${deezerTrack.id}`,
      name: deezerTrack.title,
      artists: [{
        id: deezerTrack.artist.id,
        name: deezerTrack.artist.name,
        external_urls: { spotify: deezerTrack.link },
        uri: `deezer:artist:${deezerTrack.artist.id}`,
      }],
      album: {
        id: deezerTrack.album.id,
        name: deezerTrack.album.title,
        artists: [{
          id: deezerTrack.artist.id,
          name: deezerTrack.artist.name,
          external_urls: { spotify: deezerTrack.link },
          uri: `deezer:artist:${deezerTrack.artist.id}`,
        }],
        images: [{
          height: 300,
          width: 300,
          url: deezerTrack.album.cover_medium,
        }],
        release_date: new Date().toISOString().split('T')[0], // Fallback date
        external_urls: { spotify: deezerTrack.link },
        uri: `deezer:album:${deezerTrack.album.id}`,
      },
      duration_ms: deezerTrack.duration * 1000,
      preview_url: deezerTrack.preview, // This is the key part!
      external_urls: { spotify: deezerTrack.link },
      uri: `deezer:track:${deezerTrack.id}`,
      popularity: 50, // Default popularity
      explicit: false, // Deezer doesn't provide this info easily
    };
  }
}

export const deezerService = new DeezerService();
